### PR TITLE
Feat/#39/implement user auth

### PR DESCRIPTION
**Pull request by Bogdan-Zinovij**
Added functionality of registration, login, logout, refresh tokens and verify token using JWT

### CODE REVIEW

File: services/users/.env [hunk 1]
```diff
@@ -1 +1,24 @@
-EXPRESS_PORT=8080
\ No newline at end of file
+EXPRESS_PORT=8080
```
Comment (artemmatiushenko1) [inline services/users/.env]: Is it necessary to commit your original `.env` file if you have an `.example`?

File: services/users/.gitignore [hunk 2]
```diff
@@ -1 +1,2 @@
-node_modules
\ No newline at end of file
+node_modules
+.env
```
Comment (artemmatiushenko1) [inline services/users/.gitignore:2]: hmm, strange that the .env wasn't ignored

File: services/users/src/db/models/user.model.js [hunk 3]
```diff
@@ -1,39 +1,40 @@
...
       allowNull: false,
     },
     email: {
-      type: DataTypes.STRING(64),
+      type: DataTypes.STRING,
+      allowNull: false,
+    },
+    role: {
+      type: DataTypes.STRING,
+      defaultValue: 'user',
```
Comment (artemmatiushenko1) [inline services/users/src/db/models/user.model.js]: I guess you should create an constant like object containing all of the available roles and use it. Smth like
```js
const ROLES = {
   User: 0,
   Admin: 1
}
```

File: services/users/src/services/auth.service.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    const userDto = new UserDto(user);
+    const tokens = tokenService.generateTokens(userDto);
+    await tokenService.saveToken(userDto.id, tokens.refreshToken);
+
+    return { ...tokens, user: userDto };
+  }
+
+  async refresh(refreshToken) {
+    if (!refreshToken) throw new Error(errorMessages.NOT_AUTHORIZED);
```
Comment (artemmatiushenko1) [inline services/users/src/services/auth.service.js]: I guess `NOT_AUTHORIZED` might be renamed as `UNAUTHORIZED`

File: services/users/src/services/user.service.js [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    return user;
+  }
+
+  async getUserByEmail(email) {
+    return await User.findOne({ where: { email } });
+  }
+
+  async createUser(userData) {
+    userData.id = uuid();
```
Comment (artemmatiushenko1) [inline services/users/src/services/user.service.js]: It's a bad practice to modify function arguments, I guess you should firstly copy userData 
```js
const user = {...userData}
``` 
and then add new fields

**Review by artemmatiushenko1 — CHANGES_REQUESTED** [review]

File: services/users/src/services/user.service.js (hunk 5 above)
Comment (Bogdan-Zinovij) [reply services/users/src/services/user.service.js]: Fixed

File: services/users/src/services/auth.service.js (hunk 4 above)
Comment (Bogdan-Zinovij) [reply services/users/src/services/auth.service.js]: Fixed

File: services/users/.gitignore (hunk 2 above)
Comment (KhrapkoVasyl) [reply services/users/.gitignore:2]: To ignore it, firstly you need to delete it manually from repo

File: services/users/src/services/auth.service.js [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    const tokens = tokenService.generateTokens(userDto);
+    await tokenService.saveToken(userDto.id, tokens.refreshToken);
+
+    return { ...tokens, user: userDto };
+  }
+
+  async signOut(refreshToken) {
+    if (!refreshToken) throw new Error(errorMessages.NOT_AUTHORIZED);
+
+    return await tokenService.removeToken(refreshToken);
```
Comment (KhrapkoVasyl) [inline services/users/src/services/auth.service.js]: extra await here

File: services/users/src/services/token.service.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    return await Token.create({ userId, refreshToken });
+  }
+
+  async findToken(refreshToken) {
+    return await Token.findOne({ where: { refreshToken } });
+  }
+
+  async removeToken(refreshToken) {
+    return await Token.destroy({ where: { refreshToken } });
```
Comment (KhrapkoVasyl) [inline services/users/src/services/token.service.js]: extra await

File: services/users/src/services/token.service.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    const tokenData = await Token.findOne({ where: { userId } });
+    if (tokenData) {
+      return await Token.update({ refreshToken }, { where: { userId } });
+    }
+
+    return await Token.create({ userId, refreshToken });
+  }
+
+  async findToken(refreshToken) {
+    return await Token.findOne({ where: { refreshToken } });
```
Comment (KhrapkoVasyl) [inline services/users/src/services/token.service.js]: extra await

File: services/users/src/services/token.service.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    return { accessToken, refreshToken };
+  }
+
+  async saveToken(userId, refreshToken) {
+    const tokenData = await Token.findOne({ where: { userId } });
+    if (tokenData) {
+      return await Token.update({ refreshToken }, { where: { userId } });
+    }
+
+    return await Token.create({ userId, refreshToken });
```
Comment (KhrapkoVasyl) [inline services/users/src/services/token.service.js]: extra await

File: services/users/src/services/token.service.js [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+      expiresIn: process.env.JWT_REFRESH_EXPIRES,
+    });
+
+    return { accessToken, refreshToken };
+  }
+
+  async saveToken(userId, refreshToken) {
+    const tokenData = await Token.findOne({ where: { userId } });
+    if (tokenData) {
+      return await Token.update({ refreshToken }, { where: { userId } });
```
Comment (KhrapkoVasyl) [inline services/users/src/services/token.service.js]: extra await

File: services/users/src/services/user.service.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
+import bcrypt from 'bcryptjs';
+import { v4 as uuid } from 'uuid';
+import { User } from '../db/models/user.model.js';
+import { errorMessages, SALT } from '../config.js';
+
+class UserService {
+  async getUsers() {
+    return await User.scope('withoutPassword').findAll({ order: ['id'] });
```
Comment (KhrapkoVasyl) [inline services/users/src/services/user.service.js]: extra await

File: services/users/src/services/user.service.js [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  async getUserById(id) {
+    const user = await User.findOne({ where: { id } });
+
+    if (!user) throw new Error(errorMessages.USER_NOT_EXISTS_ID);
+
+    return user;
+  }
+
+  async getUserByEmail(email) {
+    return await User.findOne({ where: { email } });
```
Comment (KhrapkoVasyl) [inline services/users/src/services/user.service.js:20]: extra await

File: services/users/src/services/user.service.js [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  async createUser(userData) {
+    userData.id = uuid();
+    userData.password = await bcrypt.hash(userData.password, SALT);
+
+    await User.create(userData);
+
+    return await User.findOne({ where: { id: userData.id } });
```
Comment (KhrapkoVasyl) [inline services/users/src/services/user.service.js]: extra await

File: services/users/src/services/user.service.js [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+  }
+
+  async updateUser(id, userData) {
+    const user = await User.findOne({ where: { id } });
+    if (!user) throw new Error(errorMessages.USER_NOT_EXISTS_ID);
+
+    await User.update(userData, { where: { id } });
+
+    return await User.findOne({ where: { id } });
```
Comment (KhrapkoVasyl) [inline services/users/src/services/user.service.js]: extra await

**Review by KhrapkoVasyl — CHANGES_REQUESTED** [review]

File: services/users/src/db/models/user.model.js (hunk 3 above)
Comment (Bogdan-Zinovij) [reply services/users/src/db/models/user.model.js]: Fixed

File: services/users/.gitignore (hunk 2 above)
Comment (Bogdan-Zinovij) [reply services/users/.gitignore:2]: the file already existed in the remote repository before it got into gitignore. now it is fixed

**Review by artemmatiushenko1 — APPROVED** [review]

File: services/users/src/config.js [hunk 15]
```diff
@@ -1,6 +1,22 @@
-'use strict';
+export const PREFIX = '/api/v1';
+export const SALT = 7;
 
-const PREFIX = '/api/v1';
-const SALT = 7;
+export const kafkaTopics = {
+  SUCCESSFUL_REGISTRATION: 'new-user-registered',
```
Comment (KhrapkoVasyl) [inline services/users/src/config.js]: I suggest renaming it to clarify the topic name
```suggestion
  NEW_USER_REGISTERED: 'new-user-registered',
```

**Review by KhrapkoVasyl — APPROVED** [review]
One small change. Overall - good job

**Pull request merged**